apps/server/test/lexical/server/code_intelligence/completion/env_test.exs

```diff
@@ -228,12 +233,132 @@ defmodule Lexical.Server.CodeIntelligence.Completion.EnvTest do
+    end
+
+    test "is true when the struct is in the function variable" do
+      env = "%User{|}" |> func_variable() |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true when the struct is in the function arguments" do
+      env = "%User{|}" |> func_args() |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is false when the cursor is not in a container" do
+      env = new_env("%User{|}")
+      refute in_context?(env, :struct_arguments)
+    end
+
+    test "is true if the cursor after the field name" do
+      env = "%User{name: |}" |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true if the cursor after the field value" do
+      env = "%User{name: \"John\"|}" |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true if the cursor starts in the middle of the struct" do
+      env = "%User{name: \"John\", |}" |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is false if the cursor after the closing curly" do
+      env = "%User{}|" |> my_module() |> new_env()
+      refute in_context?(env, :struct_arguments)
+    end
+
+    test "is true if the cursor in current module arguments" do
+      env = "%__MODULE__{|}" |> func_variable() |> my_module() |> new_env()
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true even the struct alias is at the above line" do
+      source = ~q[
+        %User{
+          name: "John",
+          |
+        }
+      ]
+      env = new_env(source)
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true even if the value of a struct key is a tuple" do
+      env = new_env("%User{favorite_numbers: {3}|")
+      assert in_context?(env, :struct_arguments)
+    end
+
+    test "is true even if the cursor is at a nested struct" do
+      env = new_env("%User{address: %Address{}|")
+      assert in_context?(env, :struct_arguments)
+    end
+  end
+
+  describe "in_context?(env, :value)" do
+    test "is true if the cursor is after a value character" do
+      env = new_env("%{foo: 1|}")
+      assert in_context?(env, :value)
+    end
+
+    test "is ture if the cursor is after a colon" do
+      env = new_env("%{foo: |}")
+      assert in_context?(env, :value)
+    end
+
+    test "is false if the cursor is in a multiple lines key positon" do
+      source = ~q[
+        %{
+          foo: 1,
+          |
+        }
+      ]
+
+      env = new_env(source)
+      refute in_context?(env, :value)
+    end
+  end
+
   describe "in_context?(env, :struct_reference)" do
     test "is true if the reference starts on the beginning of the line" do
       env = new_env("%User|")
       assert in_context?(env, :struct_reference)
     end
 
+    test "is false if the cursor in curly" do
```

Comment:
so, here's my thinking... I actually think this should be `true`. We're in a struct reference for sure, but we're _also_ in the arguments. Thoughts?

Reply:
Semantically it is true, but from a usage point of view, struct reference is currently only used for struct completions, and it requires changes to the implementation of **struct references**.

If we need to change it, we must do it in another PR. From a practical point of view, I don't see a very strong reason at the moment

Reply:
aside from things being correctly named, and doing one thing and one thing well? To me, there are two things you're checking for, and they've been muddled together, which is why things aren't working well. They are:

1. We're inside a struct reference. 
2. We're inside map / keyword update syntax. 
When both are true, we're doing what you want. We can improve ergonomics on `in_context?` if you like, by doing something like;
```elixir
in_context?(env, [:struct_reference, :map_update])
```
Which will return true if you're in both a struct reference and a map update reference.

Reply:
> and it requires changes to the implementation of struct references.

I don't understand, why?

Reply:
I think I can remove this test because this test is for `struct_references` not `struct_arguments`, but my pr didn't change any product code of `in_context(env, :struct_references)`

I really don't want to keep this pr inflated, of course we can move this discussion to a new issue, and use another PR to solve this.